table/tables/partition_test.go

```diff
@@ -700,6 +702,1550 @@ func TestIssue31629(t *testing.T) {
+				},
+				{
+					"SELECT count(*) FROM tkey_string WHERE id8 = 'a' or id8 = 'b'",
+					false, false, true, true, []string{"partition:p1", "partition:p3"}, []string{"partition:p0", "partition:p2"}, 3,
+				},
+				{
+					"SELECT count(*) FROM tkey_string WHERE id8 > 'a' AND id8 < 'c'",
+					false, false, true, true, []string{"partition:p1", "partition:p2", "partition:p0", "partition:p3"}, []string{}, 1,
+				},
+			},
+		},
+	}
+	executePartTableCase(t, tk, testCases3, createSQL3, insertSQLS3, dropSQL3)
+}
+
+func TestKeyPartitionTableMixed(t *testing.T) {
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("create database partitiondb2")
+	defer tk.MustExec("drop database partitiondb2")
+	tk.MustExec("use partitiondb2")
+	// SHOW CREATE TABLE
+	tk.MustExec("CREATE TABLE tkey1 (col1 INT NOT NULL, col2 DATE NOT NULL,col3 INT NOT NULL, col4 INT NOT NULL, UNIQUE KEY (col3))" +
+		" PARTITION BY KEY(col3)" +
+		"(PARTITION `p0`," +
+		"PARTITION `p1`," +
+		"PARTITION `p2`," +
+		"PARTITION `p3`)")
+	tk.MustQuery("show create table tkey1").Check(testkit.Rows("tkey1 CREATE TABLE `tkey1` (\n" +
+		"  `col1` int(11) NOT NULL,\n" +
+		"  `col2` date NOT NULL,\n" +
+		"  `col3` int(11) NOT NULL,\n" +
+		"  `col4` int(11) NOT NULL,\n" +
+		"  UNIQUE KEY `col3` (`col3`)\n" +
+		") ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin\n" +
+		"PARTITION BY KEY (`col3`) PARTITIONS 4"))
+
+	// BLOB, JSON don't support key partition
+	err := tk.ExecToErr("create table tkey_string(\n" +
+		"id5 BLOB not null,\n" +
+		"id6 TEXT not null,\n" +
+		"name varchar(16)\n" +
+		") PARTITION BY KEY(id5) partitions 4\n")
+	require.Error(t, err)
+	require.Regexp(t, "Field 'id5' is of a not allowed type for this type of partitioning", err)
+
+	// BLOB, JSON don't support key partition
+	err = tk.ExecToErr("create table tkey_string2(\n" +
+		"id5 BLOB not null,\n" +
+		"id6 TEXT not null,\n" +
+		"name varchar(16)\n" +
+		") PARTITION BY KEY(id6) partitions 4\n")
+	require.Error(t, err)
+	require.Regexp(t, "Field 'id6' is of a not allowed type for this type of partitioning", err)
+
+	err = tk.ExecToErr("CREATE TABLE tkey_json (c1 JSON) PARTITION BY KEY(c1) partitions 4")
+	require.Error(t, err)
+	require.Regexp(t, "Field 'c1' is of a not allowed type for this type of partitioning", err)
+
+	// It doesn't support LINEAR KEY partition
+	tk.MustExec("CREATE TABLE tkey_linear (col1 INT, col2 CHAR(5), col3 DATE) PARTITION BY LINEAR KEY(col3) PARTITIONS 5")
+	result := tk.MustQuery("show warnings")
+	result.CheckContain("LINEAR KEY is not supported, using non-linear KEY instead")
+
+	// It will ignore ALGORITHM=1|2
+	tk.MustExec("CREATE TABLE tkey_algorithm1 (col1 INT, col2 CHAR(5), col3 DATE) PARTITION BY KEY ALGORITHM=1 (col3) PARTITIONS 5")
+	tk.MustExec("CREATE TABLE tkey_algorithm2 (col1 INT, col2 CHAR(5), col3 DATE) PARTITION BY KEY ALGORITHM=2 (col3) PARTITIONS 5")
```

Comment:
Should we issue a warning for the use of ALGORITHM ?

Reply:
um...maybe showing a warning is more resonable, but I keep it with the original process.